test/suites/expiration/suite_test.go

```diff
@@ -66,6 +68,310 @@ var _ = AfterEach(func() { env.Cleanup() })
+					Values:   []string{"2xlarge"},
+				},
+			)
+			nodePool.Spec.Disruption.Budgets = []corev1beta1.Budget{{
+				Nodes: "50%",
+			}}
+			nodePool.Spec.Disruption.ExpireAfter = corev1beta1.NillableDuration{}
+
+			var numPods int32 = 6
+			dep := coretest.Deployment(coretest.DeploymentOptions{
+				Replicas: numPods,
+				PodOptions: coretest.PodOptions{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							corev1beta1.DoNotDisruptAnnotationKey: "true",
+						},
+						Labels: map[string]string{"app": "large-app"},
+					},
+					// Each 2xlarge has 8 cpu, so each node should fit 2 pods.
+					ResourceRequirements: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							v1.ResourceCPU: resource.MustParse("3"),
+						},
+					},
+				},
+			})
+			selector := labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)
+			env.ExpectCreated(nodeClass, nodePool, dep)
+
+			env.EventuallyExpectCreatedNodeClaimCount("==", 3)
+			env.EventuallyExpectCreatedNodeCount("==", 3)
+			env.EventuallyExpectHealthyPodCount(selector, int(numPods))
+			env.Monitor.Reset() // Reset the monitor so that we can expect a single node to be spun up after expiration
+
+			nodes := env.EventuallyExpectNodeCount("==", 3)
+			By("adding finalizers to the nodes to prevent termination")
+			// Add a finalizer to each node so that we can stop termination disruptions
+			for _, node := range nodes {
+				node.Finalizers = append(node.Finalizers, common.TestingFinalizer)
+				env.ExpectUpdated(node)
+			}
+
+			By("making the nodes empty")
+			// Delete the deployment to make all nodes empty.
+			env.ExpectDeleted(dep)
+
+			By("enabling expiration")
+			nodePool.Spec.Disruption.ExpireAfter = corev1beta1.NillableDuration{Duration: lo.ToPtr(time.Second * 30)}
+			env.ExpectUpdated(nodePool)
+
+			// Expect that the NodeClaims will all be marked expired
+			Eventually(func(g Gomega) {
+				nodeClaimList := &corev1beta1.NodeClaimList{}
+				err := env.Client.List(env.Context, nodeClaimList)
+				g.Expect(err).To(Succeed())
+				lo.ForEach(nodeClaimList.Items, func(nc corev1beta1.NodeClaim, _ int) {
+					g.Expect(nc.StatusConditions().GetCondition(corev1beta1.Expired).IsTrue()).To(BeTrue())
+				})
+			}).Should(Succeed())
+
+			// Expect that two nodes are tainted.
+			nodes = env.EventuallyExpectTaintedNodeCount("==", 2)
+
+			// Add a finalizer to each node so that we can stop termination disruptions
+			for _, node := range nodes {
+				Expect(env.ExpectTestingFinalizerRemoved(node)).To(Succeed())
+			}
+
+			// After the deletion timestamp is set and all pods are drained
+			// the node should be gone
+			env.EventuallyExpectNotFound(nodes[0], nodes[1])
+
+			// Expect that only one node is tainted, even considering the new node that was just created.
+			nodes = env.EventuallyExpectTaintedNodeCount("==", 1)
+
+			// Expect the finalizers to be removed and deleted.
+			Expect(env.ExpectTestingFinalizerRemoved(nodes[0])).To(Succeed())
```

Comment:
Shouldn't need this check here right? Or was the intention for this finalizer to not be removed earlier?

Reply:
Since there were 3 total nodes, this is to allow the third node to be deleted.

Reply:
Oh I was being silly, I missed that `env.EventuallyExpectTaintedNodeCount` reassigned nodes prior to removing the finalizer before so I thought it was already removed on all 3. Looks good 👍